src/gich.rs

```diff
@@ -0,0 +1,181 @@
+/// It provides access to the GIC hypervisor interface registers through a non-null pointer.
+///
+/// # Safety
+///
+/// The base pointer must point to valid GIC hypervisor interface registers mapped in memory.
+pub struct GicHypervisorInterface {
+    base: NonNull<GicHypervisorInterfaceRegs>,
+}
+
+unsafe impl Send for GicHypervisorInterface {}
+unsafe impl Sync for GicHypervisorInterface {}
+
+/// GicHypervisorInterface provides an interface to interact with the GIC (Generic Interrupt Controller)
+/// in a hypervisor context. It allows reading and writing to various GIC registers.
+impl GicHypervisorInterface {
+    /// Construct a new GIC Hypervisor interface instance from the base address.
+    ///
+    /// # Arguments
+    ///
+    /// * `base` - A mutable pointer to the base address of the GIC registers.
+    ///
+    /// # Returns
+    ///
+    /// A new instance of `GicHypervisorInterface`.
+    pub const fn new(base: *mut u8) -> Self {
+        Self {
+            base: NonNull::new(base).unwrap().cast(),
+        }
+    }
+
+    /// Get a reference to the GIC Hypervisor Interface Registers.
+    ///
+    /// # Safety
+    ///
+    /// This function dereferences a raw pointer, so it is marked as unsafe.
+    const fn regs(&self) -> &GicHypervisorInterfaceRegs {
+        unsafe { self.base.as_ref() }
+    }
+
+    /// Get the value of the HCR (Hypervisor Control Register).
+    ///
+    /// # Returns
+    ///
+    /// The current value of the HCR register.
+    pub fn get_hcr(&self) -> u32 {
+        self.regs().HCR.get()
+    }
+
+    /// Set the value of the HCR (Hypervisor Control Register).
+    ///
+    /// # Arguments
+    ///
+    /// * `hcr` - The value to set the HCR register to.
+    pub fn set_hcr(&self, hcr: u32) {
+        self.regs().HCR.set(hcr);
+    }
+
+    /// Get the value of the ELRSR (Empty List Register Status Register) at the specified index.
+    ///
+    /// # Arguments
+    ///
+    /// * `idx` - The index of the ELRSR register to read.
+    ///
+    /// # Returns
+    ///
+    /// The current value of the specified ELRSR register.
+    pub fn elrsr(&self, idx: usize) -> u32 {
+        self.regs().ELRSR[idx].get()
+    }
+
+    /// Get the value of the EISR (Empty Interrupt Status Register) at the specified index.
+    ///
+    /// # Arguments
+    ///
+    /// * `idx` - The index of the EISR register to read.
+    ///
+    /// # Returns
+    ///
+    /// The current value of the specified EISR register.
+    pub fn eisr(&self, idx: usize) {
+        self.regs().EISR[idx].get();
+    }
+
+    /// Get the value of the MISR (Maintenance Interrupt Status Register).
+    ///
+    /// # Returns
+    ///
+    /// The current value of the MISR register.
+    pub fn misr(&self) -> u32 {
+        self.regs().MISR.get()
+    }
+
+    /// Get the value of the APR (Active Priority Register).
+    ///
+    /// # Returns
+    ///
+    /// The current value of the APR register.
+    pub fn apr(&self) -> u32 {
```

Comment:
how to use this api when you want to get if a target preemption level (0-31) is active or not?

Reply:
`GICH.apr() & 1 << priority` is a good usage.